### DISCUSSION TITLE
New class: rsSingleSineModeler

### DISCUSSION THREAD

**Discussion by RobinSchmidt** [RobinSchmidt/RS-MET#320] [Show and tell]
i'm working on a new class called `rsSingleSineModeler` - this is a sort of extension or further development of the old sine-recreation free functions. i'm putting all this functionality neatly together in a class and i've also written lots of additional code already to better handle edge cases and also a completely new algorithm to estimate the sine parameters (but it's not yet fully finished - it sometimes works well but at other times not so much). eventually, my goal is to model *any* signal as an amplitude- and frequency-modulated sinusoid - whether it is one or not. the input can even be white noise and the analysis should still produce some amplitude and phase-data that allows identity resynthesis. of course, this data may then look like noise, too - but maybe for some intermediate cases, we'll get something interesting and useful out of it. we'll see

**Comment by elanhickler** [comment]
is this going to help in phaselocking or creating perfect phase samples of acoustic instruments?

**Comment by RobinSchmidt** [comment]
well, it's a sort of version 2.0 of all this rsSineFrequency, rsSineAmplitudeAndPhase stuff, which - if i remember correctly - is what you use for phase-locking stuff. it basically takes an input signal x[n] and models it as a[n] * sin(p[n]) with instantaneous amplitude and phase array a[n],p[n] - where the instantaneous phase p[n] can optionally be further converted to instantaneous frequency with optional phase modulation - so, yes, i guess so. the process is most meaningfully applied to extracted partials......but my actual goal with it is to have fun with applying it to non-sinusoidal signals and see what sort of effects we can get with it. in your partial recreation applications, you are using rsSineFrequencyAt together with rsSineAmplitudeAndPhase right? and that can also achieve identity resynthesis, when the parameters are not modified, right? ...but this rsSineFrequencyAt function is only applicable to periodic (ideally sinusoidal) inputs

...sooo well...i guess when being applied to extracted partials, it will probably not behave very differently from what we already have since a long time. i expect the most interesting benefits to happen when it's applied to more complex signals for which the old algo was not applicable
